File: crates/varde-code/tests/fixtures/php/structural.php
<?php
// Structural fixtures: Function (function_definition / method_declaration),
// Class, Interface, Parameter, Variable (property_element), MemberAccess, plus
// Extends (base_clause) and Implements (class_interface_clause). A `trait`
// carries implementation, so it maps to the Class kind (see REQUIRED_KINDS in
// src/extract/langs/php.rs).

namespace App\Http;

interface Greeter
{
    public function greet(string $name): string;
}

trait SomeTrait
{
    public function shared()
    {
        return 1;
    }
}

class UserController extends Base implements Greeter
{
    private int $count = 0;

    public function __construct(int $count)
    {
        $this->count = $count;
    }

    public function greet(string $name): string
    {
        return $this->count;
    }
}

function helper($x, ...$rest)
{
    return $x;
}

File: crates/varde-code/tests/fixtures/ruby/structural.rb
# Structural fixtures: Function (method / singleton_method), Class, Interface
# (module), Parameter, Variable, MemberAccess, plus Extends (superclass) and
# Implements (include). Ruby has no interface keyword, so `module` maps to the
# Interface kind (see REQUIRED_KINDS in src/extract/langs/ruby.rs).

module Describable
  def describe(other)
    other
  end
end

class Greeter < Base
  include Describable

  attr_reader :greeting

  def initialize(greeting = "hi")
    @greeting = greeting
  end

  def greet(other = "world", *rest, punct:, **opts)
    message = "Hello #{other}"
    message.upcase
  end

  def self.build(greeting)
    new(greeting)
  end
end
